### PR TITLE
Opt einsum

### PR DESCRIPTION
**Pull request by Thenerdstation**
This uses opt_einsum's contractor order code to find the best contraction order for a tensor network.

### CODE REVIEW

**Comment by Thenerdstation** [comment]
https://github.com/google/TensorNetwork/issues/171

File: tensornetwork/contractors/opt_einsum_paths/optimal_path.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import opt_einsum
+import tensornetwork
+from tensornetwork.contractors.opt_einsum_paths import utils
+
+
+def optimal(net: tensornetwork.TensorNetwork) -> tensornetwork.TensorNetwork:
+  """Optimal contraction order via opt_einsum.
+
+  This method will find the truely optimal contraction order via
```
Comment (sleichen) [inline tensornetwork/contractors/opt_einsum_paths/optimal_path.py]: truly

**Comment by Thenerdstation** [comment]
> Do you have any benchmark data indicating how this implementation performs relative to naive?

Not yet, but I could. Is there a good TN to use? Preferably one with large bond dimension and few nodes. Any ideas?

**Comment by Thenerdstation** [comment]
```python
import tensornetwork
from tensornetwork.contractors import naive, optimal
import numpy as np
import timeit

def build_mps_network(n=4, d=32):
  net = tensornetwork.TensorNetwork(backend="numpy")
  dangling_edges = []
  nodes = []
  for i in range(n):
    nodes.append(net.add_node(np.ones((d, d, d))))
    dangling_edges.append(nodes[-1][2])
  for a, b in zip(nodes, nodes[1:]):
    a[0] ^ b[1]
  return net, dangling_edges


def benchmark_ladder_network(contractor):
  net1, edges1 = build_mps_network()
  net2, edges2 = build_mps_network()
  net1.add_subnetwork(net2)
  for a, b in zip(edges1, edges2):
    a ^ b
  return contractor(net1).get_final_node()


print("Naive: ", timeit.timeit(
  lambda: benchmark_ladder_network(naive), number=1))

print("Optimal: ", timeit.timeit(
  lambda: benchmark_ladder_network(naive), number=1))
```

Prints out
```
Naive:  45.7
Optimal:  34.2
```

Most of the time in `optimal` was spent just finding the contraction order based on core usage. Actual contraction time was ~3 seconds.

**Comment by Thenerdstation** [comment]
Just hit a bug. Let me fix that first...

**Comment by Thenerdstation** [comment]
Fixed it 
```python
import tensornetwork
from tensornetwork.contractors import naive, optimal
import numpy as np
import timeit

def build_mps_network(n=4, d=24):
  net = tensornetwork.TensorNetwork(backend="numpy")
  dangling_edges = []
  nodes = []
  for i in range(n):
    nodes.append(net.add_node(np.ones((d, d, d))))
    dangling_edges.append(nodes[-1][2])
  for a, b in zip(nodes, nodes[1:]):
    a[0] ^ b[1]
  return net, dangling_edges


def benchmark_ladder_network(contractor):
  net1, edges1 = build_mps_network()
  net2, edges2 = build_mps_network()
  net1.add_subnetwork(net2)
  for a, b in zip(edges1, edges2):
    a ^ b
  return contractor(net1).get_final_node()


print("Naive: ", timeit.timeit(
  lambda: benchmark_ladder_network(naive), number=1))

print("Optimal: ", timeit.timeit(
  lambda: benchmark_ladder_network(optimal), number=1))
```

This prints
```
Naive:  5.176
Optimal:  0.075
```

So a huge improvement.

**Pull request merged**